File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys


def dfs(a, n):
    global min_value
    if a == (N // 2):
        start, link = 0, 0
        for i in range(N):
            for j in range(N):
                if visited[i] and visited[j]:
                    start += S[i][j]
                elif not visited[i] and not visited[j]:
                    link += S[i][j]
        min_value = min(min_value, abs(start - link))
        return
    else:
        for x in range(n, N):
            visited[x] = True
            dfs(a + 1, x + 1)
            visited[x] = False


N = int(sys.stdin.readline())
S = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]
visited = [False] * N
min_value = 2_000
dfs(0, 0)
print(min_value)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <iostream>
#define MAX 20
using namespace std;

int n, minDiff = 100;
int s[MAX][MAX];
bool visited[MAX];

void dfs(int depth, int idx) {
	if (depth == n / 2) {
		int start = 0, link = 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (visited[i] && visited[j]) start += s[i][j];
				else if (!visited[i] && !visited[j]) link += s[i][j];
			}
		}
		int diff = abs(start - link);
		if (diff < minDiff) minDiff = diff;
		return;
	}
	for (int i = idx; i < n; i++) {
		visited[i] = true;
		dfs(depth + 1, i + 1);
		visited[i] = false;
	}
}

int main() {
	cin.tie(0)->sync_with_stdio(0);

	cin >> n;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> s[i][j];
		}
	}
	dfs(0, 0);
	cout << minDiff;
}